doc/aaa/TACACS+ Requirement.md

```diff
@@ -0,0 +1,161 @@
+# SONiC TACACS+ Protocol Requirement
+
+# Table of Contents
+- [Table of Contents](#table-of-contents)
+- [About this Manual](#about-this-manual)
+- [1 Functional Requirements](#1-functional-requirement)
+  * [1.1 Authentication](#11-authentication)
+  * [1.2 Authorization](#12-authorization)
+  * [1.3 Accounting](#13-accounting)
+  * [1.4 User script](#14-user-script)
+  * [1.5 Docker support](#15-docker-support)
+  * [1.6 Multiple TACAS server](#16-multiple-tacas-server)
+- [2 Configuration and Management Requirements](#2-configuration-and-management-requirements)
+  * [2.1 SONiC CLI](#21-sonic-cli)
+  * [2.2 Config DB](#22-config-db)
+  * [2.3 Counter](#23-counter)
+  * [2.4 System log](#24-system-log)
+- [3 Limitation](#limitation)
+  * [3.1 Command size](#31-command-size)
+  * [3.2 Server count](#32-server-count)
+
+# About this Manual
+This document provides a detailed description on the requirement of TACACS+ protocol support.
+
+# 1 Functional Requirement
+## 1.1 Authentication
+- Authentication when user login to SONiC host.
+
+
+## 1.2 Authorization
+- Authorization when:
+	- User login to SONiC host.
+	- User run command on SONiC host.
+		- User can only run commands in whitelist.
+		- For recursive command/script, only the top level command have authorization.
+
+- Only command in whitelist  visible to user:
+	- All privilege level share same whitelist to simplify design.
+	- Different privilege level have different permission to run these command.
+	- All commands in sudoers will add to the whitelist. and sudoers config file still need for RO users, this is because when remote TACACS server not avaliable, we need use local group permission for failover.
+
+- Disable user behavior in shell:
+	- Changing directories with the cd builtin.
+		- User still can access files in other folder, and use 'ls' command to list content in other folder, for example:
+        ```
+            test@testhost:~$ ls -l /etc/
+            total 880
+            drwxr-xr-x 3 test test       4096 Aug 22  2020 NetworkManager
+            drwxr-xr-x 7 test test       4096 Aug 22  2020 X11
+            drwxr-xr-x 3 test test       4096 Aug 22  2020 acpi
+        ```
+	- Setting or unsetting the values of the SHELL, PATH, HISTFILE, ENV, or BASH_ENV variables.
+	- Specifying command names containing slashes.
+	- Importing function definitions from the shell environment at startup.
+	- Parsing the value of SHELLOPTS from the shell environment at startup.
+	- Redirecting output using the ‘>’, ‘>|’, ‘<>’, ‘>&’, ‘&>’, and ‘>>’ redirection operators.
```

Comment:
>Redirecting [](http://example.com/codeflow?start=3&length=11)

What are the alternatives for user to achieve similar goal?

Reply:
By patch bash, we can allow user use redirect, will remove this from document.

Reply:
Do you have a link to the patch? I am wondering if redirect could be configurable. For example, RW users could use redirect as normal, but RO users are banned to use redirect.

Reply:
It's a very simple code change, remove following code in  redir.c line:836

	#if defined (RESTRICTED_SHELL)
	      if (restricted && (WRITE_REDIRECT (ri)))
	{
	  free (redirectee_word);
	  return (RESTRICTED_REDIRECT);
	}
	#endif /* RESTRICTED_SHELL */

So for RW allow and RO disable, we can:
1. create a variable isReadonlyUser
2. When bash startup, check current user group and set this variable.
3. If any behavior we want to disable for RO only we can change code like this:

	#if defined (RESTRICTED_SHELL)
	      if (restricted && **isReadonlyUser** && (WRITE_REDIRECT (ri)))
	{
	  free (redirectee_word);
	  return (RESTRICTED_REDIRECT);
	}
	#endif /* RESTRICTED_SHELL */